armi/reactor/geometry.py

```diff
@@ -113,6 +123,308 @@ def __str__(self):
+        """
+        if isinstance(source, BoundaryType):
+            return source
+        elif isinstance(source, str):
+            return cls.fromStr(source)
+        else:
+            raise TypeError("Expected str or BoundaryType; got {}".format(type(source)))
+
+    @classmethod
+    def fromStr(cls, symmetryStr: str) -> "BoundaryType":
+        # case-insensitive
+        canonical = symmetryStr.lower().strip()
+        if canonical == NO_SYMMETRY:
+            return cls.NO_SYMMETRY
+        elif canonical == PERIODIC:
+            return cls.PERIODIC
+        elif canonical == REFLECTIVE:
+            return cls.REFLECTIVE
+
+        errorMsg = (
+            "{} is not a valid boundary option. Valid boundary options are:".format(
+                str(canonical)
+            )
+        )
+        errorMsg += ", ".join([f"{sym}" for sym in boundaryTypes])
+        raise ValueError(errorMsg)
+
+    @property
+    def label(self):
+        """Human-presentable label"""
+
+        if self == self.NO_SYMMETRY:
+            return "No Symmetry"
+        elif self == self.REFLECTIVE:
+            return "Reflective"
+        elif self == self.PERIODIC:
+            return "Periodic"
+
+    def __str__(self):
+        """Inverse of fromStr()"""
+        if self == self.NO_SYMMETRY:
+            return ""
+        elif self == self.PERIODIC:
+            return PERIODIC
+        elif self == self.REFLECTIVE:
+            return REFLECTIVE
+
+    def hasSymmetry(self):
+        return not self == self.NO_SYMMETRY
+
+
+class SymmetryType:
+    """
+    A thin wrapper for DomainType and BoundaryType enumerations.
+
+    (see GeomType class for explanation of why we want enumerations for DomainType and
+    BoundaryType).
+
+    The goal of this class is to provide simple functions for storing these options
+    in enumerations and using them to check symmetry conditions, while also providing
+    a standard string representation of the options that facilitates interfacing with
+    yaml and/or the database nicely.
+    """
+
+    VALID_SYMMETRY = {
+        (DomainType.FULL_CORE, BoundaryType.NO_SYMMETRY, False),
+        (DomainType.FULL_CORE, BoundaryType.NO_SYMMETRY, True),
+        (DomainType.THIRD_CORE, BoundaryType.PERIODIC, False),
+        (DomainType.QUARTER_CORE, BoundaryType.PERIODIC, False),
+        (DomainType.QUARTER_CORE, BoundaryType.REFLECTIVE, False),
+        (DomainType.QUARTER_CORE, BoundaryType.PERIODIC, True),
+        (DomainType.QUARTER_CORE, BoundaryType.REFLECTIVE, True),
+        (DomainType.EIGHTH_CORE, BoundaryType.PERIODIC, False),
+        (DomainType.EIGHTH_CORE, BoundaryType.REFLECTIVE, False),
+        (DomainType.EIGHTH_CORE, BoundaryType.PERIODIC, True),
+        (DomainType.EIGHTH_CORE, BoundaryType.REFLECTIVE, True),
+        (DomainType.SIXTEENTH_CORE, BoundaryType.PERIODIC, False),
+        (DomainType.SIXTEENTH_CORE, BoundaryType.REFLECTIVE, False),
+    }
```

Comment:
I think a dict of lists is a good solution. Should I put the check against that dict into the function `checkValidGeomSymmetryCombo` to log the error when it's not valid? Or just drop it altogether and check the dict directly?

Reply:
I think a function still makes sense. The underlying data structure is kind of specific and not necessarily something that we want to guarantee as part of the module's interface. The same argument could be made for VALID_SYMMETRY above. Might be a good idea to hide the list of tuples, since even the reason _why_ they are tuples in the first place is peculiar to the implementation.

Reply:
Honestly if anything is using VALID_SYMMETRY outside of this module (e.g. that voluptuous stuff, where i recommended making it do the set comprehension itself), the VALID_SYMMETRY that it sees should probably contain instances of SymmetryType, rather than tuples. Maybe make the tuple version private and derive the SymmetryType instances in the public VALID_SYMMETRY collection from those?

Reply:
It was easy enough to eliminate VALID_SYMMETRY from the other code, so it's only used inside of SymmetryType now. So we don't need to construct a whole list of SymmetryType instances from the valid options.